algebra/ring.lean

```diff
@@ -57,6 +57,151 @@ section comm_ring
+λ hy, have h : y * x ∈ S, from mul_mem hy, by rwa [mul_comm]
+
+end is_ideal
+
+@[reducible] def zero_ideal (α : Type u) [comm_ring α] : set α := {(0:α)}
+instance zero_ideal.is_ideal (α : Type u) [comm_ring α] : is_ideal α $ zero_ideal α :=
+by refine {..}; intros; simp [set.mem_singleton_iff] at *; simp [*]
+
+@[reducible] def univ_ideal (α : Type u) [comm_ring α] : set α := set.univ
+instance univ_ideal.is_ideal (α : Type u) [comm_ring α] : is_ideal α $ univ_ideal α :=
+by refine {..}; intros; trivial
+
+theorem is_ideal.eq_univ_of_contains_unit {α : Type u} [comm_ring α] (S : set α) [is_ideal α S] :
+(∃ x:α, x ∈ S ∧ is_unit x) → S = set.univ :=
+λ ⟨x, hx, y, hy⟩, set.ext $ λ z, ⟨λ hz, trivial, λ hz, calc
+   z = (x * y) * z : by simp [hy]
+ ... = x * (y * z) : mul_assoc x y z
+ ... ∈ S : is_ideal.mul_mem hx⟩
+
+theorem is_ideal.univ_of_one_mem {α : Type u} [comm_ring α] (S : set α) [is_ideal α S] :
+(1:α) ∈ S → S = set.univ :=
+λ h, set.ext $ λ z, ⟨λ hz, trivial, λ hz, by simpa using (is_ideal.mul_mem h : 1 * z ∈ S)⟩
+
+instance is_ideal.hom_preimage {α : Type u} {β : Type v} [comm_ring α] [comm_ring β]
+(f : α → β) [is_hom f] (S : set β) [is_ideal β S] : is_ideal α (f ⁻¹' S) :=
+{ zero_mem := by simp [is_hom.map_zero f],
+  add_mem  := λ x y (hx : f x ∈ S) hy, by simp [is_hom.map_add f, is_ideal.add_mem hx hy],
+  mul_mem  := λ x y (hx : f x ∈ S), by simp [is_hom.map_mul f, is_ideal.mul_mem hx] }
+
+class is_prime_ideal {α : Type u} [comm_ring α] (S : set α) extends is_ideal α S : Prop :=
+(ne_univ_ideal : S ≠ univ_ideal α)
+(mem_or_mem_of_mul_mem : ∀ {x y : α}, x * y ∈ S → x ∈ S ∨ y ∈ S)
+
+theorem mem_or_mem_of_mul_eq_zero {α : Type u} [comm_ring α] (S : set α) [is_prime_ideal S] :
+  ∀ {x y : α}, x * y = 0 → x ∈ S ∨ y ∈ S :=
+λ x y hxy, have x * y ∈ S, by rw hxy; from is_ideal.zero_mem S,
+is_prime_ideal.mem_or_mem_of_mul_mem this
+
+class is_maximal_ideal {α : Type u} [comm_ring α] (S : set α) extends is_ideal α S : Prop :=
+mk' ::
+  (ne_univ_ideal : S ≠ univ_ideal α)
+  (eq_or_univ_of_subset : ∀ (T : set α) [is_ideal α T], S ⊆ T → T = S ∨ T = univ_ideal α)
+
+theorem is_maximal_ideal.mk {α : Type u} [comm_ring α] (S : set α) [is_ideal α S] :
+  (1:α) ∉ S → (∀ x (T : set α) [is_ideal α T], S ⊆ T → x ∉ S → x ∈ T → (1:α) ∈ T) → is_maximal_ideal S :=
+λ h₁ h₂,
+{ _inst_2 with
+  ne_univ_ideal := λ hu, have (1:α) ∈ S, by rw hu; trivial, h₁ this,
+  eq_or_univ_of_subset := λ T ht hst, or.cases_on (classical.em $ ∃ x, x ∉ S ∧ x ∈ T)
+    (λ ⟨x, hxns, hxt⟩, or.inr $ @@is_ideal.univ_of_one_mem _ T ht $ @@h₂ x T ht hst hxns hxt)
+    (λ hnts, or.inl $ set.ext $ λ x,
+       ⟨λ hxt, classical.by_contradiction $ λ hxns, hnts ⟨x, hxns, hxt⟩,
+        λ hxs, hst hxs⟩) }
+
+theorem not_unit_of_mem_maximal_ideal {α : Type u} [comm_ring α] (S : set α) [is_maximal_ideal S] : S ⊆ nonunits α :=
+λ x hx hxy, is_maximal_ideal.ne_univ_ideal S $ is_ideal.eq_univ_of_contains_unit S ⟨x, hx, hxy⟩
+
+class local_ring (α : Type u) [comm_ring α] :=
```

Comment:
Some of this is getting a bit too theoretical for the `algebra` section. Move maximal and prime ideals, and local rings, to `ring_theory` as well.